src/gui_common/ModalManager.cs

```diff
@@ -0,0 +1,189 @@
+
+#pragma warning disable CA2213 // Disposable fields should be disposed
+    private CanvasLayer canvasLayer = null!;
+    private Control activeModalContainer = null!;
+#pragma warning restore CA2213 // Disposable fields should be disposed
+
+    private System.Collections.Generic.Dictionary<CustomWindow, Node> parents = new();
+    private Stack<CustomWindow> modalStack = new();
+
+    private bool modalsDirty = true;
+
+    private ModalManager()
+    {
+        instance = this;
+
+        PauseMode = PauseModeEnum.Process;
+    }
+
+    public static ModalManager Instance => instance ?? throw new InstanceNotLoadedYetException();
+
+    /// <summary>
+    ///   Checks whether the top-most modal on the modal stack is an exclusive popup.
+    /// </summary>
+    public bool IsAnyExclusivePopupActive => GetCurrentlyActiveExclusivePopup() != null;
+
+    public override void _Ready()
+    {
+        canvasLayer = GetNode<CanvasLayer>("CanvasLayer");
+        activeModalContainer = canvasLayer.GetNode<Control>("ActiveModalContainer");
+
+        activeModalContainer.SetAsToplevel(true);
+    }
+
+    public override void _Process(float delta)
+    {
+        if (modalsDirty)
+        {
+            UpdateModals();
+            modalsDirty = false;
+        }
+    }
+
+    /// <summary>
+    ///   Promotes the given popup as a modal and make it visible.
+    /// </summary>
+    /// <remarks>
+    ///   <para>
+    ///     This does Node re-parenting operation, therefore calling this plainly in <see cref="Node._Ready"/> wouldn't
+    ///     work as the Node is still busy then. Alternatively, you could defer the call in the next frame by using
+    ///     <see cref="Invoke.Queue"/>.
+    ///   </para>
+    /// </remarks>
+    public void MakeModal(CustomWindow popup)
+    {
+        if (modalStack.Contains(popup))
+            return;
+
+        parents[popup] = popup.GetParent();
+        modalStack.Push(popup);
+        modalsDirty = true;
+
+        var bind = new Array { popup };
+        popup.CheckAndConnect("hide", this, nameof(OnModalLost), bind, (uint)ConnectFlags.Oneshot);
+
+        if (!popup.Visible)
+            popup.Open();
+    }
+
+    /// <summary>
+    ///   Closes any currently active modal popups.
+    /// </summary>
+    [RunOnKeyDown("ui_cancel", Priority = Constants.POPUP_CANCEL_PRIORITY)]
+    public bool HideCurrentlyActivePopup()
+    {
+        if (modalStack.Count <= 0)
+            return false;
+
+        var popup = modalStack.Peek();
+
+        if (popup.Exclusive && !popup.ExclusiveAllowCloseOnEscape)
+            return false;
+
+        popup.Close();
+        popup.Notification(Control.NotificationModalClose);
+
+        if (popup is CustomDialog dialog)
+            dialog.EmitSignal(nameof(CustomDialog.Cancelled));
+
+        return true;
+    }
+
+    /// <summary>
+    ///   Returns the top-most exclusive popup in the current Viewport's modal stack. Null if there is none.
```

Comment:
This documentation is wrong, and I don't know how to fix it... this only checks if the topmost item in the stack is an exclusive popup, it doesn't look for any exclusive popups in the stack. The same problem applies to the `IsAnyExclusivePopupActive` property, as that name is really bad now. It doesn't check for *any* exclusive popups, just if there happens to be an exclusive popup on top...

Reply:
Well, depends on how you interpret it. The "active" part in the member names refer to the _top-most_ exclusive popup. The only thing that's in fact wrong in that documentation is that it still mentions Viewport, I guess I forgot to change that.

Reply:
All my English language skills tell me that, that's not the correct interpretation, "top-most exclusive popup" is the "top-most" popup that is exclusive. I.e. the logic or the documentation (and the variable name) has to change as they don't match. The code "returns the top most popup if it is exclusive", whereas the documentation "returns the top most popup that is exclusive". The slight difference in wording makes them mean completely different things.

Reply:
Sure, my suggestion for better names would be `GetTopMostPopupIfExclusive` for the method and `IsTopMostPopupExclusive` for the property, the documentations can be slightly adjusted accordingly.

Reply:
Those sound good to me. I assume there's a reason why only the top-most popup being exclusive matters.

Reply:
Previously, there was no need to know whether the topmost popup be otherwise. Apart from that, no other reason really.

Reply:
I suppose that is fine. I was thinking that there could be a bug / non-optimal situation where a non-exclusive popup gets opened on top of an exclusive one, or there might be a really tricky stack of popups like:
- Non-exclusive (top)
- Exclusive
- Non-exclusive
- Exclusive

The situation where that could trigger should be very rare, but it would be nice if the system would handle it in *some* logical way.

Reply:
If you mean that there would be a problem with hiding popups with ESC due to that then I think there shouldn't, inputs should always be consumed (unless you intentionally made it otherwise) by the topmost popup. Lower popups whether it be exclusive or non-exclusive, shouldn't be affected in any way.

The point is, exclusive popups and the above discussed class members does _not_ affect popup hides with ESC.